File: 프로그래머스/lv2/김원래/프로세스.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

struct compare
{
    bool operator()(int& a, int& b)
    {
        return a < b;
    }
};

priority_queue<int, vector<int>, compare> pq;
queue<pair<int,int>> prioritiesQ;

int solution(vector<int> priorities, int location) {
    int answer = 0;
    
    int counter = 0;
    
    for(int priority : priorities) {
        prioritiesQ.push({priority, counter++});
        pq.push(priority);
    }
    
    int pos = 1;
    while(!prioritiesQ.empty()) {
        pair<int,int> qFront = prioritiesQ.front();
        prioritiesQ.pop();
        
        int pqTop = pq.top();
        
        if(qFront.first < pqTop) {
            prioritiesQ.push(qFront);
        }
        else {
          pq.pop();
        
            if(qFront.second == location) {
                answer = pos;
                break;
            }
            
            ++pos;
        }
        
        
        
    }
    
    return answer;
}

File: 프로그래머스/lv2/이선영/프로세스.java
import java.util.*;

class Solution {
    public int solution(int[] priorities, int location) {
        int answer = 0;
        PriorityQueue<Integer> pq= new PriorityQueue<>((o1,o2)-> o2-o1);
        
        for(int i=0; i<priorities.length; i++){
            pq.offer(priorities[i]);
        }
        
        Queue<int[]> q = new LinkedList<>();
        
        for(int i=0; i<priorities.length; i++){
            q.offer(new int[]{i, priorities[i]});
        }
        
        while(!q.isEmpty()){
            
            if(q.peek()[1] == pq.peek()){ //우선순위가 같을때
                if(q.peek()[0] == location)
                    return ++answer; //location이 같을때
                else{
                    q.poll();
                    pq.poll();
                    answer++;
                }
            }else{
                int[] poll = q.poll();
                q.offer(poll);
                
            }
            
            
            
        }
        
        return answer;
    }
}
